---
title: "scPagwas"
date: "Last Updated: `r format(Sys.time(), '%d, %B, %Y at %H:%M')`"
output:
  md_document:
    variant: markdown_github
editor_options:
  chunk_output_type: console
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# About scPagwas

**scPagwas** employs a polygenic regression model to prioritize a set of trait-relevant genes and uncover trait-relevant cell subpopulations by incorporating pathway activity transformed single-cell RNA sequencing (scRNA-seq) data with genome-wide association studies (GWAS) summary data.

<img src="./man/figures/Figure1.1.png" width="100%" style="display: block; margin: auto;" />
 
The methodology and benchmarking performance are described in: 

Polygenic regression uncovers trait-relevant cellular contexts through pathway activation transformation of single-cell RNA sequence data.(2023) 


Code for reproducing the analysis from the paper is available [here](https://github.com/dengchunyu/scPagwas_reproduce). 

For further usage on the scPagwas package, you can visit the [tutorials](https://dengchunyu.github.io/scPagwas/) or [website](https://dengchunyu.github.io/about/). A vignette for using also can be accessed using browseVignettes("scPagwas")

## Installation

You can install the released version of scPagwas from [github](https://github.com/sulab-wmu/scPagwas) with: 

``` r
#install some dependence packages
install.packages("SeuratObject")
install.packages("Seurat")
install.packages("ggpubr")
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("GenomicRanges")
BiocManager::install("IRanges")

devtools::install_github("sulab-wmu/scPagwas")
```

## Usage 
quick-start example: 
```{r message=FALSE, eval = FALSE}
 library(scPagwas)

 #1.start to run the wrapper functions for example.
 Pagwas_data<-scPagwas_main(Pagwas = NULL,
                     gwas_data =system.file("extdata", "GWAS_summ_example.txt", package = "scPagwas"), # The GWAS Summary statistics files 
                     Single_data =system.file("extdata", "scRNAexample.rds", package = "scPagwas"),# scRNA-seq data in seruat format with "RNA" assays and normalized.
                     output.prefix="test", # the prefix name for output files
                     output.dirs="scPagwastest_output",# the directory file's name for output
                     block_annotation = block_annotation,# gene position in chromosome is provided by package.
                     assay="RNA", # the assays for scRNA-seq data to use.
                     Pathway_list=Genes_by_pathway_kegg,# pathway list is provided by package, including gene symbols.
                     chrom_ld = chrom_ld,# The LD data is provided by package.
                     singlecell=T, # Whether to run the singlecell process.
                     celltype=T# Whether to run the celltype process.
)

```

## Tutorials

scPagwas provides a number of tutorials for various situation.

- The [Data_input_preproccess_for_scPagwas](https://dengchunyu.github.io/dataprepare/2023/03/10/Data_input_preproccess_for_scPagwas.html) tutorial provides the methods of data-input preproccess for scPagwas.

- The [The-common-usage-and-result-explaination-for-scPagwas](https://dengchunyu.github.io/routineuse/2023/03/10/The-common-usage-and-result-explaination-for-scPagwas.html) tutorial provides the usual procedure for scPagwas including the result interpretation are discussed, and visualizing their characteristics.

- The [Running_scPagwas_steps_by_SubFunctions](https://dengchunyu.github.io/routineuse/2023/03/10/Running-scPagwas-steps-by-SubFunctions.html) tutorial provides the procedure for only cell types or single cell functions; Otherwise, a step by step introduction for scPgawas sub-functions is also provided.

- The [Multi_TraitFiles_for_one_SingleData_running](https://dengchunyu.github.io/flexibleuse/2023/03/13/Multi-Traits-for-one-Single-cell-Data.html) tutorial provides the procedure that running scPagwas based on multiple trait files in one scRNA-seq dataset.

- The [Split_big_scRNAseqData_and_integrate_result](https://dengchunyu.github.io/flexibleuse/2023/03/13/Split-big-scRNAseqData-and-integrate_result.html) tutorial provides the procedure that running scPagwas with several splited scRNA-seq datasets while the whole dataset is too big to run.

- The [Prune-GwasSummary-FIles-by-plink](https://dengchunyu.github.io/dataprepare/2023/03/13/Prune-GwasSummary-FIles-by-plink.html)tutorial provides the procedure that prune the big gwas summary files.
