---
title: "The common usage and result explaination for scPagwas"
date: "Last Updated: `r format(Sys.time(), '%d, %B, %Y at %H:%M')`"
output: rmarkdown::html_vignette
vignette: |
  %\VignetteIndexEntry{Bmmc_monocytecount_singlecell_celltype_vignette} 
  %\VignetteEncoding{UTF-8} 
  %\VignetteEngine{knitr::rmarkdown}
---

```{r, include = FALSE}
library(knitr)
knitr::opts_chunk$set(
 collapse = TRUE,
 comment = "#>",
  fig.path = "figures/vignette-",
 out.width = "60%"
)
```

## Preprocessed data
This is a real example for the artical. We used BMMC scRNA-seq data and monocyte count trait to test scPagwas. 
The processed monocytecount gwas data can be download from [here](https://1drv.ms/t/s!As-aKqXDnDUHi6sx7Hqblj2Sgl7P8w?e=cb5Ihf).

BMMC example scRNA-seq data can be obtained from [here](https://1drv.ms/u/s!As-aKqXDnDUHjfcMjN4VGtAw1Utm0w?e=ocUrEx).

## 1. Compute the singlecell and celltype result for monocytecount trait

If we running scPagwas in multi-core in server environment, there may cause an error: `Error: Two levels of parallelism are used. See ?assert_cores\`\` add this code before call in R environment:

```{bash eval = FALSE}
export OPENBLAS_NUM_THREADS=1
```

There is no need to run this code in window system.

In this example, we run the scPagwas for usual step, both running singlecell and celltype process.

```{r eval = FALSE}
library(scPagwas)
Pagwas<-scPagwas_main(Pagwas =NULL,
                     gwas_data="monocytecount_prune_gwas_data.txt",
                     Single_data ="Seu_Hema_data.rds",
                     output.prefix="monocytecount_scPagwas",
                     output.dirs="monocytecount_bmmc",
                     Pathway_list=Genes_by_pathway_kegg,
                     n.cores=2,
                     assay="RNA",
                     singlecell=T, 
                     celltype=T,
                     block_annotation = block_annotation,
                     chrom_ld = chrom_ld)
```

Sometimes, we need to remove the objects in cache folder: 

```{r eval = FALSE}
if(length(SOAR::Objects())>0){
 SOAR::Remove(SOAR::Objects()) 
}
```
## 2. Result interpretation

There are two types of result, Seruat format return result and files output; The running time is relatively long, we can download the result files from [here]()

### 2.1 Pagwas

```{r eval = FALSE}
> print(Pagwas)
An object of class Seurat 
16488 features across 35582 samples within 3 assays 
Active assay: RNA (15860 features, 5000 variable features)
 2 other assays present: scPagwasPaHeritability, scPagwasPaPca
 3 dimensional reductions calculated: pca, tsne, umap

> names(Pagwas@misc)
[1] "Pathway_list"                  "pca_cell_df"                  
[3] "lm_results"                    "bootstrap_results"            
[5] "scPathways_rankPvalue"         "gene_heritability_correlation"
```


1.  In this Seruat result, two new assays were added:

    -   **scPagwasPaPca**: An assay for S4 type of data; the svd score result for pathways in each cells;

    -   **scPagwasPaHeritability**: An assay for S4 type of data; the gPas score matrix for pathways in each cells;

<!-- -->

2.  In meta.data, four columns were added:

    -   **scPagwas.TRS.Score1**: inheritance related score, enrichment socre for genetics top genes;

    -   **scPagwas.gPAS.score**: Inheritance pathway regression effects score for each cells;

    -   **CellpValue**: Ranked CellpValue for each cells;

    -   **CellqValue**: Ranked CellqValue for each cells, adjust p value.

<!-- -->

3.  A new element names misc is added in scPagwas's result, `Pagwas@misc` is a list including:

    -   **Pathway_list**: filtered pathway gene list;

    -   **pca_cell_df**: a pahtway and celltype data frame for pathway svd(1'st pca) result for each celltype;

    -   **lm_results**: the regression result for each celltype;

    -   **gene_heritability_correlation**: heritability correlation value for each gene;

    -   **bootstrap_results**: The bootstrap data frame results for celltypes including bootstrap pvalue and confidence interval.

### 2.1 Pagwas : output files result

<img src="D:\OneDrive\RPakage\scPagwas\vignettes\figures\resultfiles.png" width="50%" />
In **monocytecount_bmmc**  folder, There is several result files :

-   **\*\_cellytpes_bootstrap_results.csv** : The result of cellytpes for bootstrap p results;

-    **\*\_gene_heritability_correlation.csv** : The result of gene heritability correlation with gPAS score.

-   **\*\_Pathway_singlecell_lm_results.txt** : The regression result for all pahtway and single cell matrix;

-   **\*\_singlecell_Pathways_scGene_ranlP.csv** : The ranked pvalue for eache singlecell;

-   **\*\_singlecell_scPagwas_score_pvalue.Result.csv** : The data frame result for each cell inlcuding scPagwas.TRS.Score, scPagwas.gPAS.score, pValueHighScale, qValueHighScale;

**scPagwas_cache** is a temporary folder to save the SOAR data, you can remove it when finish the scPagwas.

## 3. Result Visualization

### 3.1 Visualize the scPagwas Score of single cell data in UMAP or TSNE plot.

1) DimPlot of singlecell data.

```{r eval = FALSE}
 require("RColorBrewer")
 require("Seurat")
 require("SeuratObject")
 require("ggsci")
 require("dplyr")
 require("ggplot2")
 require("ggpubr")
 #check the objects
color26 <- c("#D9DD6B","#ECEFA4","#D54C4C","#8D2828","#FDD2BF","#E98580","#DF5E5E","#492F10","#334257","#476072","#548CA8","#00A19D","#ECD662","#5D8233","#284E78","#3E215D","#835151","#F08FC0","#C6B4CE","#BB8760","#FFDADA","#3C5186","#558776","#E99497","#FFBD9B","#0A1D37")

Seurat::DimPlot(Pagwas,pt.size=1,reduction="tsne",label = T, repel=TRUE)+
 scale_colour_manual(name = "celltypes", values = color26)+
 umap_theme()+ggtitle("Monocyte BMMC")+labs(x="TSNE",y="")+theme(aspect.ratio=1)


```

<img src="D:\OneDrive\RPakage\scPagwas\vignettes\figures\monocyte_bmmc_dimplot_umap.png" width="50%" />

scPagwas.TRS.Score1 and positive(p<0.05) cells showing in dimplot.

```{r eval = FALSE}
 scPagwas_Visualization(Single_data=Pagwas,
                        p_thre = 0.05,
                        FigureType = "tsne",
                        width = 7,
                        height = 7,
                        lowColor = "white", 
                        highColor = "red",
                        output.dirs="figure",
                        size = 0.5,
                        do_plot = T)
```

2) scPagwas.gPAS.score dimplot. 

<img src="D:\OneDrive\RPakage\scPagwas\vignettes\figures\scPagwas.gPAS.score_tsne.png" width="40%" />

3) scPagwas.TRS.Score dimplot. 

<img src="D:\OneDrive\RPakage\scPagwas\vignettes\figures\scPagwas.TRS.Score_tsne.png" width="40%" />

4) scPagwas_CellScaleqValue dimplot. 

<img src="D:\OneDrive\RPakage\scPagwas\vignettes\figures\scPagwas_CellScaleqValue0.05_tsne.png" width="30%" />

Positive cells(q value<0.05): red dot; Negative cells: other dot. 

### 3.2 Plot the barplot of the proportion of positive Cell. 

Plot the barplot of the proportion of positive Cells in celltypes:
```{r eval = FALSE}
plot_bar_positie_nagtive(seurat_obj=Pagwas,
                         var_ident="celltypes",
                         var_group="positiveCells",
                         vec_group_colors=c("#E8D0B3","#7EB5A6"),
                         do_plot = T)
```

<img src="D:\OneDrive\RPakage\scPagwas\vignettes\figures\plot_bar_positie_nagtive.png" width="60%" />

Plot the barplot of the proportion of celltypes in positive and negative Cells:

```{r eval = FALSE}
plot_bar_positie_nagtive(seurat_obj=Pagwas,
                              var_ident="positiveCells",
                              var_group="celltypes",
                              p_thre = 0.01,
                              vec_group_colors=NULL,
                              f_color=colorRampPalette(brewer.pal(n=10, name="RdYlBu")),
                              do_plot = T)

```

<img src="D:\OneDrive\RPakage\scPagwas\vignettes\figures\plot_bar_positie_nagtive2.png" width="60%" />


### 3.3 Plot the heritability correlated genes

```{r eval = FALSE}
heritability_cor_scatterplot(gene_heri_cor=Pagwas@misc$gene_heritability_correlation,
                             topn_genes_label=10,
                             color_low="#035397",
                             color_high ="#F32424",
                             color_mid = "white",
                             text_size=2,
                             do_plot=T,
                             max.overlaps =20,
                             width = 7,
                             height = 7)

```

<img src="D:\OneDrive\RPakage\scPagwas\vignettes\figures\heritability_cor_scatterplot.png" width="50%" />

### 3.4 Show expression of the top heritability correlation genes in celltypes

```{r eval = FALSE}
library("ggsci")
library("Seurat")
top5genes<-rownames(Pagwas@misc$gene_heritability_correlation)[order(Pagwas@misc$gene_heritability_correlation,decreasing = T)[1:5]]
'plot_vln_Corgenes(seurat_obj=Pagwas,
             assay="RNA", slot="data",
             var_group="celltypes",
             vec_features=top5genes,
             vec_group_colors= color26,
             do_plot = T
             )'
```

<img src="D:\OneDrive\RPakage\scPagwas\vignettes\figures\plot_vln_Corgenes.png" width="60%" />


### 3.5 celltypes bootstrap_results reuslt 
Barplot for celltypes 

```{r eval = FALSE}
Bootstrap_P_Barplot(p_results=Pagwas@misc$bootstrap_results$bp_value[-1],
                    p_names=rownames(Pagwas@misc$bootstrap_results)[-1],
                    figurenames = "Bootstrap_P_Barplot.pdf",
                    width = 5,
                    height = 7,
                    do_plot=T,
                    title = "monocytecount_bmmc")
```

<img src="D:\OneDrive\RPakage\scPagwas\vignettes\figures\Bootstrap_P_Barplot.png" width="40%" />

Forest plot for estimate value. 
```{r eval = FALSE}
Bootstrap_estimate_Plot(Pagwas=Pagwas,
                        width = 9,
                        height = 7,
                        do_plot=T)

```

<img src="D:\OneDrive\RPakage\scPagwas\vignettes\figures\estimate_Plot_00.png" width="60%" />

```{r}
sessionInfo()
```
