---
title: "Split_big_scRNAseqData_and_integrate_result"
date: "Last Updated: `r format(Sys.time(), '%d, %B, %Y at %H:%M')`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Split_big_scRNAseqData_and_integrate_result}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
library(knitr)
knitr::opts_chunk$set(
 collapse = TRUE,
 comment = "#>",
  fig.path = "figures/vignette-",
 out.width = "60%"
)
```

```{r setup}
library(scPagwas)
```

## Run the scPagwas for big scRNA-seq data 
`If your computer storage is enough, ignore this item`.  
Sometimes, the storage is limit for your analysis, you can splice the scRNA-seq data set by celltypes or in order. 


### 1. Splice scRNA-seq data 
Note, you can split the scRNA-seq data by celltypes or in random.
When you split the scRNA-seq data in random, you should set the 
`min_clustercells=1` 

```{r eval = FALSE}
library(Seurat)
scRNAexample <-readRDS(system.file("extdata", "scRNAexample.rds", package = "scPagwas"))
#set the output.dirs for any directory.
set.dirs="D:/OneDrive/GWAS_Multiomics/Pagwas/test/"
#set the number of split time, it depend on the size of your single cell data.
#There have two form of spliting the data:

#########
###1.By Random
#Create the random index number.
n_split=2
Split_index <- rep(1:n_split, time = ceiling(ncol(scRNAexample)/n_split), length = ncol(scRNAexample))

for (i in 1:n_split) {
  Example_splice <- scRNAexample[,Split_index==i]
  saveRDS(Example_splice,file = paste0(set.dirs,"Example_splice",i,".rds"))
}

```

### 2. Run gwas data input 

Second, we run the `GWAS_summary_input` and `Snp2Gene` firstly, because the following function share the same gwas data.
Sometimes, we need not run these functions for small gwas data, because a little time for these functions.

```{r eval = FALSE}
Pagwas<-list()
gwas_data <- bigreadr::fread2(system.file("extdata", "GWAS_summ_example.txt", package = "scPagwas"))
Pagwas <- GWAS_summary_input(
    Pagwas = Pagwas,
    gwas_data = gwas_data
  )
Pagwas$snp_gene_df <- SnpToGene(gwas_data = Pagwas$gwas_data, 
                                block_annotation = block_annotation, marg = 10000)
names(Pagwas)

```

### 2. Run scPagwas for sub-data

```{r eval = FALSE}

setwd(set.dirs)

for (i in 1:n_split) {
 scPagwas_main(Pagwas =Pagwas,
                     gwas_data =NULL,
                     Single_data = paste0(set.dirs,"Example_splice",i,".rds"), #the 
                     output.prefix=i, #the prefix can be the circulation coefficient i.
                     output.dirs="splice_scPagwas", #the output.dirs shoukd be the same for each circulation
                     Pathway_list=Genes_by_pathway_kegg,
                     run_split=TRUE, #You must set the key parameter.This parameter is set to run single result for each split result.
                     min_clustercells=10, #the minimum cluster cell number should be 1.
                     assay="RNA",
                     block_annotation = block_annotation,
                     chrom_ld = chrom_ld)
  gc()
}

```


### 3. Integrate the result for splice data. 

We integrate the splice result by `merge_pagwas` function. 

```{r eval = FALSE}
setwd(set.dirs) #you must set the workfiles the same as before
Pagwas_integrate <- merge_pagwas(Single_data = system.file("extdata", "scRNAexample.rds", package = "scPagwas") ,# read the whole single cell data.
                                 output.dirs="splice_scPagwas", #The same with scPagwas_main functionn_topgenes = 1000,
                         assay='RNA',
                         random=T,
                         seed=1234,
                         random_times=5,
                         proportion=0.5)

```
